### PR TITLE
Server Status Ping API

### PR DESCRIPTION
**Pull request by Minecrell**
Adds an API for #331 to the SpongeAPI. This implements most things mentioned there except direct access to protocol versions, instead there is a (currently) read-only wrapper around them as `GameVersion`.

I decided to use the newer names from the protocol rewrite in 1.7 instead of the older names from the legacy ping. That's why "server list ping" is "status ping" and "MotD" is "description" now. This matches these names to the ones used in the protocol and the Minecraft source, and should also prevent confusion about "MotD" in the server list, or "MotD" as the message if you join a server (like used by Essentials) in the future.

### CODE REVIEW

**Comment by dobrakmato** [comment]
Is **Favicon** the right name for the class? Wouldn't the **ServerIcon** or just **Icon** be better?

**Comment by Minecrell** [comment]
@dobrakmato Favicon is the named used in the protocol and in Minecraft internally. It's the name I have always used for this, but I can change it if others agree that it should be named differently.

File: src/main/java/org/spongepowered/api/GameRegistry.java [hunk 1]
```diff
@@ -383,4 +389,44 @@
...
+
+    /**
+     * Loads a favicon from a specified {@link File}.
+     *
+     * @param file The favicon file
+     * @return The loaded favicon from the file
+     * @throws IOException If the favicon couldn't be loaded
+     * @throws FileNotFoundException If the file doesn't exist
+     */
+    Favicon loadFavicon(File file) throws IOException;
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: `throws IOException, FileNotFoundException`

File: src/main/java/org/spongepowered/api/Server.java [hunk 2]
```diff
@@ -197,9 +203,11 @@
     boolean getOnlineMode();
 
     /**
-     * Gets the message that is displayed in the server list of the client.
-     * @return The servers MOTD
+     * Gets the default message that is displayed in the server list of the
+     * client.
+     * 
+     * @return The servers default MotD
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/Server.java]: `@return The server's default Motd`

File: src/main/java/org/spongepowered/api/entity/player/User.java [hunk 3]
```diff
@@ -36,14 +36,7 @@
...
  * This is in contrast to Player which represents the ingame entity associated with an online User.
  */
-public interface User extends Identifiable, ArmorEquipable, DataSerializable {
-
-    /**
-     * Gets the player's last known username.
-     * 
-     * @return The player's last known username
-     */
-    String getName();
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/entity/player/User.java]: Why? Seems out of scope

File: src/main/java/org/spongepowered/api/status/StatusClient.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     * Returns whether the client is using an older version that doesn't support
+     * all of the features in {@link StatusResponse}.
+     * <p>
+     * For Vanilla, this returns {@code true} for all clients older than 1.7.
+     * </p>
+     *
+     * @return {@code True} if the client is using an older version
+     */
+    boolean isLegacy();
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/status/StatusClient.java]: Might also be useful to have a method for if the StatusClient is a future version, because I remember that's also a possible error.

In addition, the error given when this happens is 'outdated', not 'legacy' client

**Comment by gratimax** [comment]
This seems mostly okay, but I would also try to consider the 'streaming favicon' use case, a-la ServerListPlus :wink:. How would one implement that in this API? Could they?

File: src/main/java/org/spongepowered/api/status/StatusClient.java (hunk 4 above)
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/status/StatusClient.java]: No. That's not what this method does. It won't return false for 1.7 clients. It returns if the client is older than 1.7 (before the Netty rewrite) when the features were more limited. Although I guess I could add these methods..

**Comment by Minecrell** [comment]
@gratimax I have considered that, but as explained in https://github.com/SpongePowered/SpongeAPI/issues/331#issuecomment-68214105, it's basically just a bug on the client that is making this possible, there is nothing in the protocol that would allow this. The client is just not validating the packets sent by the server, so by resending the status response packet again and again it will keep the connection open infinitely.

This is not hard to fix on the client and requires quite some modifications on the server (and has some other problems on the client caused by this), so this is not something that should be in the SpongeAPI. However, it should be possible for mods to extend this API and add new functionality.

File: src/main/java/org/spongepowered/api/GameRegistry.java (hunk 1 above)
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: This is already included in `IOException` (`FileNotFoundException` is a subclass of `IOException`) and IntelliJ IDEA doesn't stop with complaining that it is already defined.

**Comment by Aaron1011** [comment]
The precise effect of cancelling this event should be documented, as a Vanilla client always expects a response.

**Comment by Minecrell** [comment]
@Aaron1011 Added.
